alibi_detect/cd/context_aware.py

```diff
@@ -22,8 +23,10 @@ def __init__(
             preprocess_x_ref: bool = True,
             update_ref: Optional[Dict[str, int]] = None,
             preprocess_fn: Optional[Callable] = None,
-            x_kernel: Callable = None,
-            c_kernel: Callable = None,
+            # x_kernel: Callable = None,
+            x_kernel: BaseKernel = None,
+            # c_kernel: Callable = None,
+            c_kernel: BaseKernel = None,
```

Comment:
Have we fully discussed the implications of relying on the `BaseKernel` interface here? I suppose `Callable` is the most flexible thing for users as they can literally do anything, but it also doesn't specify the interface enough (i.e. need to take two parameters of a certain type and return something of a certain type). As far as I can see this change basically asks the following of a user wanting a custom kernel:
1. Subclass from `BaseKernel`
2. Implement `forward` according to the interface

This approach is ok, but perhaps we could even be less restrictive and only ask that a user specified kernel implements a `forward` method with the right types (this could be implemented using `typing.Protocol`). Of course this would not be enough if internally we rely on certain kernel attributes in which case we need to stick with the subclassing approach. @mauicv @ascillitoe

Reply:
As a follow-up, I see we rely on things like `kernel.trainable` checks which would rule out the less restrictive approach, just want to make sure we're fully happy with that.

Reply:
In fact the restriction to subclassing is even stronger than I thought because we're implicitly saying that kernels can only be either `nn.Module` or `tf.keras.Model` types (and coupled the kernel interface to the backend fairly strongly). 

Understand it stems mostly from the use of trainable kernels and using these restrictive interfaces makes it easier to deal with in the code (e.g. can always assume certain attributes exist), but could we be getting too restrictive? @arnaudvl @ojcobb

Reply:
Why does relying on kernel attributes restrict us here? I think protocols also work for [attributes](https://mypy.readthedocs.io/en/stable/protocols.html#using-isinstance-with-protocols)? The user would have to specify the attributes to match the protocol though.

Reply:
Didn't know that protocols support attributes.

Even so, would we have to be careful about the capabilities? E.g. device management via kernel attributes may not be relevant for user defined custom kernels?

Reply:
Stating the obvious perhaps, but I think we need to think quite carefully about how a user might create their own custom subclassed kernel. For example, having abstract methods etc to make sure important methods are implemented (and error messages are clear), and thinking about what attributes the user needs to consider when building a custom kernel. 

Part of this might just be making sure we have good documentation around custom kernels, also `ABC`'s or `Protocol`'s could help...

On a side note, if we have a wider range of kernels and also support elementary operations to combine kernels, are we certain that we even need the capability for fully customised kernels?

Reply:
The code here still needs cleaning up (see commented out sections). I believe we are fine though on being fairly strict and imposing the `BaseKernel`?

Reply:
Actually it seems like my above comment is not correct and `nn.Module` is the required restriction since e.g. SumKernel or ProductKernel do not inherit from BaseKernel?

Reply:
Looks like the above needs more thought, allowing `nn.Module` basically removes all advantages of having `BaseKernel`. Not sure if product/sum kernels should inherit also from `BaseKernel` (or perhaps they should be of some type `CombinedKernel` and both base+combined are allowed in the detectors?).